### PR TITLE
Dobe/dockeys

### PR DESCRIPTION
**Pull request by dobe**

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/where/HasColumn.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import io.crate.planner.symbol.DefaultTraversalSymbolVisitor;
+import io.crate.planner.symbol.Function;
+import io.crate.planner.symbol.Reference;
+import io.crate.planner.symbol.Symbol;
+
+import javax.annotation.Nullable;
+
+public class HasColumn {
+
+    public static final Visitor VISITOR = new Visitor();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/HasColumn.java]: can you change the VISITOR to private?

File: sql/src/main/java/io/crate/analyze/DeleteStatementAnalyzer.java [hunk 2]
```diff
@@ -34,14 +37,24 @@
 @Singleton
 public class DeleteStatementAnalyzer extends DefaultTraversalVisitor<AnalyzedStatement, Analysis> {
 
+    public static final String VERSION_SEARCH_EX_MSG =
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/DeleteStatementAnalyzer.java]: could also be made private

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -138,12 +137,10 @@ public Symbol convert(Expression expression, ExpressionAnalysisContext expressio
...
         if (whereExpression.isPresent()) {
-            WhereClause whereClause = new WhereClause(normalize(convert(whereExpression.get(), context)), null, null, null);
-            if(whereClause.hasQuery()){
-                WhereClauseValidator whereClauseValidator = new WhereClauseValidator();
-                whereClauseValidator.validate(whereClause);
-            }
-            return tableRelation.resolve(whereClause);
+            WhereClause whereClause = new WhereClause(normalize(convert(whereExpression.get(), context)), null, null);
+            whereClause = tableRelation.resolve(whereClause);
+            //WhereClauseValidator.validate(whereClause);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: can you remove the WhereClauseValidator here?

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -438,26 +438,42 @@ protected Symbol visitInPredicate(InPredicate node, ExpressionAnalysisContext co
                 return Literal.NULL;
             }
 
-            Set<Object> rightValues = new HashSet<>();
+            InListExpression listExpression = ((InListExpression) node.getValueList());
+
+            if (listExpression.getValues().isEmpty()) {
+                // TODO:check if this is possible
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: TODO?

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -438,26 +438,42 @@ protected Symbol visitInPredicate(InPredicate node, ExpressionAnalysisContext co
...
-                Literal rightLiteral;
-                try {
-                    rightLiteral = Literal.convert(right, leftType);
-                    rightValues.add(rightLiteral.value());
-                } catch (IllegalArgumentException | ClassCastException e) {
-                    throw new IllegalArgumentException(
-                            String.format(Locale.ENGLISH, "invalid IN LIST value %s. expected type '%s'",
-                                    SymbolFormatter.format(right),
-                                    leftType.getName()));
+                if (right.valueType() != leftType){
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: could it be that we have arrays here? in that case .equals must be used instead of identity check

Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: maybe this whole thing could be replaced by using ``castIfNeededOrFail``

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 6]
```diff
@@ -438,26 +438,42 @@ protected Symbol visitInPredicate(InPredicate node, ExpressionAnalysisContext co
...
+                        right = context.allocateFunction(CastFunctionResolver.functionInfo(right.valueType(), leftType),
+                                Arrays.asList(right));
+                    } else {
+                        throw new IllegalArgumentException(
+                                String.format(Locale.ENGLISH, "invalid IN LIST value %s. expected type '%s'",
+                                        SymbolFormatter.format(right),
+                                        leftType.getName()));
+                    }
                 }
+                FunctionIdent ident = new FunctionIdent(EqOperator.NAME, Arrays.asList(leftType, right.valueType()));
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: since the rightType must always match the left type the FunctionIdent and FunctionInfo (futher below) could be created once outside of the loop and be re-used

File: sql/src/main/java/io/crate/analyze/where/DocKeys.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    public class DocKey {
+
+        private final List<Symbol> key;
+
+        public DocKey(int pos) {
+            key = docKeys.get(pos);
+        }
+
+        @Nullable
+        public Optional<Long> version() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/DocKeys.java]: Nullable Optional?

File: sql/src/main/java/io/crate/analyze/where/DocKeys.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            if (withVersions && key.get(width) != null) {
+                return Optional.of(((Literal<Long>) key.get(width)).value());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/DocKeys.java]: if you write it like this: ``((Long) ((Literal) key.get(width)).value())`` you can avoid the Unchecked Cast warning

File: sql/src/main/java/io/crate/analyze/where/DocKeys.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+                @Nullable
+                @Override
+                public BytesRef apply(@Nullable Integer input) {
+                    return ValueSymbolVisitor.BYTES_REF.process(key.get(input));
+                }
+            });
+            return Optional.of(values);
+        }
+
+        public String id() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/DocKeys.java]: Should we cache the result here? Calling id() multiple times might be a bit expensive otherwise.

File: sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+        EqualityExtractor.ProxyInjectingVisitor.Context context =
+                new EqualityExtractor.ProxyInjectingVisitor.Context(columns, exact);
+        Symbol proxiedTree = piv.process(symbol, context);
+
+        // bail out if we have any unknown part in the tree
+        if (context.exact && context.seenUnknown){
+            return null;
+        }
+
+        //proxiedTree = normalizer.normalize(proxiedTree);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java]: commented code - can probably be removed?

File: sql/src/main/java/io/crate/analyze/where/HasColumn.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+    public static boolean appliesTo(@Nullable Symbol input, ColumnIdent columnIdent) {
+        if (input != null){
+            return VISITOR.process(input, columnIdent);
+        }
+        return false;
+    }
+
+    private static class Visitor extends DefaultTraversalSymbolVisitor<ColumnIdent, Boolean> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/HasColumn.java]: does it make sense to extend from DefaultTraversal.. if nothing from that class is used?

File: sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+
+    public List<List<Symbol>> extractParentMatches(List<ColumnIdent> columns, Symbol symbol){
+        return extractMatches(columns, symbol, false);
+    }
+    public List<List<Symbol>> extractExactMatches(List<ColumnIdent> columns, Symbol symbol) {
+        return extractMatches(columns, symbol, true);
+    }
+
+    private List<List<Symbol>> extractMatches(Collection<ColumnIdent> columns, Symbol symbol, boolean exact){
+        EqualityExtractor.ProxyInjectingVisitor piv = new EqualityExtractor.ProxyInjectingVisitor();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java]: the Visitor doesn't have a state, so it can probably be made static/final ?

File: sql/src/main/java/io/crate/analyze/where/WhereClauseAnalyzer.java [hunk 13]
```diff
@@ -21,115 +21,120 @@
...
-                whereClause.version(ctx.version());
-                if (ctx.keyLiterals() != null) {
-                    processPrimaryKeyLiterals(
-                            ctx.keyLiterals(),
-                            whereClauseContext
-                    );
-                    whereClause.primaryKeys(whereClauseContext.primaryKeys());
+        Set<Symbol> clusteredBy = null;
+        if (whereClause.hasQuery() && !tableInfo.schemaInfo().systemSchema()){
+            // TODO: add functionality of the whereclaue validator into analyzer or planner
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/WhereClauseAnalyzer.java]: is this TODO still a TODO?

File: sql/src/main/java/io/crate/operation/operator/CmpOperator.java [hunk 14]
```diff
@@ -48,10 +52,19 @@ public Symbol normalizeSymbol(Function symbol) {
         if (left.symbolType().isValueSymbol() && right.symbolType().isValueSymbol()) {
             // must be true due to the function registration (argument DataType signature)
 
-            if (compare(((Literal) left).compareTo((Literal)right))) {
-                return Literal.newLiteral(true);
+            boolean equal = false;
+            try {
+                equal = compare(((Literal) left).compareTo((Literal) right));
+            } catch (ClassCastException e){
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/operator/CmpOperator.java]: In which case can the ClassCastException happen here? CmpOperator should only be used if both types match?

File: sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java [hunk 15]
```diff
@@ -66,4 +70,16 @@ public static FunctionInfo functionInfo(DataType dataType, DataType returnType)
         }
         return new FunctionInfo(new FunctionIdent(functionName, ImmutableList.of(dataType)), returnType);
     }
+
+    @Nullable
+    public static Symbol castSymbol(Symbol symbol, DataType targetType) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java]: this is pretty similar to ``cast(..)`` in the ExpressionAnalyzer. Can one or the other be used instead?

File: sql/src/main/java/io/crate/planner/Planner.java [hunk 16]
```diff
@@ -394,13 +394,8 @@ public Plan visitSetStatement(SetAnalyzedStatement analysis, Context context) {
...
 
     private void createESDeleteNode(TableInfo tableInfo, WhereClause whereClause, IterablePlan plan) {
-        assert whereClause.primaryKeys().get().size() == 1;
-        Id pk = Iterables.get(whereClause.primaryKeys().get(), 0);
-        plan.add(new ESDeleteNode(
-                indices(tableInfo, whereClause)[0],
-                pk.stringValue(),
-                pk.routingValue(),
-                whereClause.version()));
+        DocKeys.DocKey docKey = whereClause.docKeys().get().getOnlyKey();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/Planner.java]: docKey here is unused?

File: sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java [hunk 17]
```diff
@@ -66,4 +70,16 @@ public static FunctionInfo functionInfo(DataType dataType, DataType returnType)
...
+        if (symbol.valueType() == targetType) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java]: probably need to use .equals here so that it works for array types

File: sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryConsumer.java [hunk 18]
```diff
@@ -28,6 +28,12 @@
 import io.crate.analyze.relations.AnalyzedRelation;
 import io.crate.analyze.relations.AnalyzedRelationVisitor;
 import io.crate.analyze.relations.PlannedAnalyzedRelation;
+import io.crate.analyze.relations.TableRelation;
+import io.crate.exceptions.VersionInvalidException;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.Functions;
+import io.crate.metadata.Routing;
+import io.crate.metadata.table.TableInfo;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/consumer/InsertFromSubQueryConsumer.java]: unused imports?

File: sql/src/main/java/io/crate/planner/symbol/DefaultTraversalSymbolVisitor.java [hunk 19]
```diff
@@ -23,6 +23,13 @@
 
 public abstract class DefaultTraversalSymbolVisitor<C, R> extends SymbolVisitor<C, R> {
 
+    /**
+     * calls process with a null context
+     */
+    public R process(Symbol symbol){
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/symbol/DefaultTraversalSymbolVisitor.java]: imho this could also be part of the SymbolVisitor

File: sql/src/main/java/io/crate/planner/symbol/Literal.java [hunk 20]
```diff
@@ -22,8 +22,8 @@
     protected DataType type;
 
     public final static Literal<Void> NULL = new Literal<>(DataTypes.UNDEFINED, null);
-    private final static Literal<Boolean> BOOLEAN_TRUE = new Literal<>(DataTypes.BOOLEAN, true);
-    private final static Literal<Boolean> BOOLEAN_FALSE = new Literal<>(DataTypes.BOOLEAN, false);
+    public final static Literal<Boolean> BOOLEAN_TRUE = new Literal<>(DataTypes.BOOLEAN, true);
+    public final static Literal<Boolean> BOOLEAN_FALSE = new Literal<>(DataTypes.BOOLEAN, false);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/symbol/Literal.java]: the ``Literal.newLiteral(Boolen boolValue)`` is now kind of redundant. Do we need both?

File: sql/src/test/java/io/crate/analyze/BaseAnalyzerTest.java [hunk 21]
```diff
@@ -309,7 +308,8 @@ protected AnalyzedStatement analyze(String statement) {
     }
 
     protected AnalyzedStatement analyze(String statement, Object[] params) {
-        return analyzer.analyze(SqlParser.createStatement(statement), params, new Object[0][]).analyzedStatement();
+        AnalyzedStatement stmt = analyzer.analyze(SqlParser.createStatement(statement), params, new Object[0][]).analyzedStatement();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/BaseAnalyzerTest.java]: why did you change this?

File: sql/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java [hunk 22]
```diff
@@ -0,0 +1,358 @@
...
+
+
+    @Test
+    public void testPK2_Eq1() throws Exception {
+        Symbol query = Eq(Ref("x"), Literal.newLiteral(1));
+        List<List<Symbol>> matches = analyzeExactXY(query);
+        assertNull(matches);
+    }
+
+//    private void printComparisons(Set<List<EqualityExtractor.EqProxy>> cp) {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/where/EqualityExtractorTest.java]: is this still needed?

File: sql/src/test/java/io/crate/analyze/where/WhereClauseAnalyzerTest.java [hunk 23]
```diff
@@ -152,14 +172,20 @@ private UpdateAnalyzedStatement analyzeUpdate(String stmt) {
...
-        QueriedTable sourceRelation = (QueriedTable) statement.relation();
-        TableRelation tableRelation = sourceRelation.tableRelation();
-        TableInfo tableInfo = tableRelation.tableInfo();
-        WhereClauseAnalyzer whereClauseAnalyzer = new WhereClauseAnalyzer(ctxMetaData, tableRelation);
-        return whereClauseAnalyzer.analyze(statement.relation().querySpec().where());
+                SqlParser.createStatement(stmt), args, new Object[0][]).analyzedStatement();
+        return statement.relation().querySpec().where();
+        //QueriedTable sourceRelation = (QueriedTable) statement.relation();
+//
+//
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/where/WhereClauseAnalyzerTest.java]: still needed?

File: sql/src/test/java/io/crate/analyze/where/WhereClauseAnalyzerTest.java [hunk 24]
```diff
@@ -191,14 +222,31 @@ public void testWherePartitionedByColumn() throws Exception {
...
+
+
+
+
+
+    @Test
+    public void testWherePartitionedByColumn() throws Exception {
+        DeleteAnalyzedStatement statement = analyzeDelete("delete from parted where date = 1395874800000");
+        TableRelation tableRelation = statement.analyzedRelation();
+        //WhereClauseAnalyzer whereClauseAnalyzer = new WhereClauseAnalyzer(ctxMetaData, tableRelation);
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/where/WhereClauseAnalyzerTest.java]: remove please if not needed

File: sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        @Override
+        public DataType valueType() {
+            return current.valueType();
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            throw new NotImplementedException();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java]: use UnsupportedOperation instead. NotImplementedException results in warnings (internal proprietary API)

File: sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            throw new NotImplementedException();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/EqualityExtractor.java]: use UnsupportedOperation instead. NotImplementedException results in warnings (internal proprietary API)

**Comment by mfussenegger** [comment]
occ tests fail on travis

File: sql/src/test/java/io/crate/analyze/BaseAnalyzerTest.java (hunk 21 above)
Comment (dobe) [inline sql/src/test/java/io/crate/analyze/BaseAnalyzerTest.java]: for better debugging

File: sql/src/main/java/io/crate/analyze/where/DocKeys.java [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+                            out.writeBytesRef(ValueSymbolVisitor.BYTES_REF.process(key.get(clusteredByIdx)));
+                        }
+                        for (int i = 0; i < width; i++) {
+                            if (i != clusteredByIdx) {
+                                out.writeBytesRef(ValueSymbolVisitor.BYTES_REF.process(key.get(i)));
+                            }
+                        }
+                        out.close();
+                    } catch (IOException e) {
+                        //
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/where/DocKeys.java]: maybe bubble the Exception as RuntimeException using ``Throwables.propagate()`` ?

**Comment by mfussenegger** [comment]
ack
